File: src/AqiClock.Domain/Entities/Announcement.cs
namespace AqiClock.Domain.Entities;

public enum AudienceType { Everyone, Teachers, Graduates, Am, Pm, SpecificClass }
public enum UpdateType { General, ClassStarts, Naseehah, MonthlyProgramme, YearlyProgramme }
public enum AnnouncementStatus { Draft, Scheduled, Published }

/// <summary>An audience-aware admin broadcast. Deleted announcements remain available to admin history.</summary>
public sealed record Announcement(
    Guid Id,
    string Title,
    string Body,
    DateTimeOffset CreatedAt,
    Guid CreatedBy,
    DateTimeOffset? ExpiresAt,
    AudienceType AudienceType = AudienceType.Everyone,
    Guid? AudienceClassId = null,
    UpdateType UpdateType = UpdateType.General,
    DateTimeOffset? PublishAt = null,
    string? EMasjidLink = null,
    AnnouncementStatus Status = AnnouncementStatus.Published,
    DateTimeOffset? DeletedAt = null)
{
    public bool IsExpiredAt(DateTimeOffset instant) => ExpiresAt is { } expiry && expiry <= instant;
    public bool IsPublishedAt(DateTimeOffset instant) =>
        DeletedAt is null && Status != AnnouncementStatus.Draft && (PublishAt is null || PublishAt <= instant);
    public bool CanPublishFromHistory => DeletedAt is null;
}

File: mobile/src/domain/announcement.ts
import {
  AnnouncementAudience,
  DeviceAudience,
  matchesAnnouncement,
} from './audience';

export type AnnouncementUpdateType =
  | 'general'
  | 'class_starts'
  | 'naseehah'
  | 'monthly_programme'
  | 'yearly_programme';

export type Announcement = {
  id: string;
  title: string;
  body: string;
  createdAt: string;
  createdBy: string;
  expiresAt: string | null;
  audienceType: AnnouncementAudience;
  audienceClassId: string | null;
  updateType: AnnouncementUpdateType;
  publishAt: string | null;
  eMasjidLink: string | null;
  status: string;
  deletedAt: string | null;
  isRead: boolean;
};

export function isAnnouncementVisible(
  item: Announcement,
  audience: DeviceAudience,
  now: Date,
): boolean {
  const instant = now.getTime();
  return (
    item.deletedAt == null &&
    item.status !== 'draft' &&
    (item.publishAt == null || new Date(item.publishAt).getTime() <= instant) &&
    (item.expiresAt == null || new Date(item.expiresAt).getTime() > instant) &&
    matchesAnnouncement(audience, item)
  );
}
